react/features/video-layout/functions.js

```diff
@@ -35,30 +35,49 @@ export function getMaxColumnCount() {
 }
 
 /**
- * Returns the cell count dimensions for tile view. Tile view tries to uphold
- * equal count of tiles for height and width, until maxColumn is reached in
- * which rows will be added but no more columns.
+ * Returns the cell count dimensions for tile view. Tile view tries to fill
+ * the available space optimally
  *
  * @param {Object} state - The redux store state.
- * @param {number} maxColumns - The maximum number of columns that can be
- * displayed.
  * @returns {Object} An object is return with the desired number of columns,
  * rows, and visible rows (the rest should overflow) for the tile view layout.
  */
-export function getTileViewGridDimensions(state: Object, maxColumns: number = getMaxColumnCount()) {
+export function getTileViewGridDimensions(state: Object) {
     // When in tile view mode, we must discount ourselves (the local participant) because our
     // tile is not visible.
     const { iAmRecorder } = state['features/base/config'];
     const numberOfParticipants = state['features/base/participants'].length - (iAmRecorder ? 1 : 0);
+    const { clientHeight, clientWidth } = state['features/base/responsive-ui'];
+    const aspectRatio = state['features/base/responsive-ui'].aspectRatio;
 
-    const columnsToMaintainASquare = Math.ceil(Math.sqrt(numberOfParticipants));
-    const columns = Math.min(columnsToMaintainASquare, maxColumns);
-    const rows = Math.ceil(numberOfParticipants / columns);
-    const visibleRows = Math.min(maxColumns, rows);
+    let bestColumns = 0;
+    let bestRows = 0;
+    let bestTileWidth = 0;
+    let bestFound = false;
+
+    while (!bestFound) {
+        const columns = bestColumns + 1;
+        const straightTileWidth = Math.floor(clientWidth / columns);
+
+        const rows = Math.ceil(numberOfParticipants / columns);
+        const straightTileHeight = Math.floor(clientHeight / rows);
+
+        const constrainedTileWidth = Math.ceil(straightTileHeight * aspectRatio);
+
+        const tileWidth = Math.min(straightTileWidth, constrainedTileWidth);
+
+        if (tileWidth > bestTileWidth) {
+            bestColumns = columns;
+            bestRows = rows;
+            bestTileWidth = tileWidth;
+        } else {
+            bestFound = true;
+        }
+    }
 
     return {
-        columns,
-        visibleRows
+        bestColumns,
```

Comment:
Changing the keys of the returned object surely breaks the code using it.

Reply:
Aren't these just the variable names that are inserted into the returned array? I am not that familiar with js, but this doesn't look like a dictionary or mapping of some kind.

Reply:
That's indeed a dictionary / object literal. Notice the curly braces.

Reply:
Ah. I was not aware of this shorthand property name thing. Annoying language construct, obscuring the common syntactical signature of a dictionary, so I didn't recognize it as such. I will fix.